config/base/derive/src/lib.rs

```diff
@@ -0,0 +1,685 @@
+
+/// Derive `iroha_config_base::reader::ReadConfig` trait.
+///
+/// Example:
+///
+/// ```
+/// use iroha_config_base_derive::ReadConfig;
+///
+/// #[derive(ReadConfig)]
+/// struct Config {
+///   #[config(default, env = "FOO")]
+///   foo: bool,
+///   #[config(nested)]
+///   nested: Nested
+/// }
+///
+/// #[derive(ReadConfig)]
+/// struct Nested {
+///   #[config(default = "42")]
+///   foo: u64
+/// }
+/// ```
+///
+/// Supported field attributes:
+///
+/// - `env = "<env var name>"` - read parameter from env (bound: `T: FromEnvStr`)
+/// - `env_custom` - for a type that implements `CustomEnvRead`. Conflicts with `env`.
+/// - `default` - fallback to default value (bound: `T: Default`)
+/// - `default = "<expr>"` - fallback to a default value specified as an expression
+/// - `nested` - delegates further reading (bound: `T: ReadConfig`).
+///   It uses the field name as a namespace. Conflicts with others.
+///
+/// A bound of `T: Deserialize` is required unless `env_only` is set.
+///
+/// Supported field shapes (if `nested` is not specified):
+///
+/// - `T` - required parameter
+/// - `WithOrigin<T>` - required parameter with origin data
+/// - `Option<T>` - optional parameter
+/// - `Option<WithOrigin<T>>` - optional parameter with origin data
+///
+/// If `default` attr is specified, it expects a default value for the actual field type, i.e. `T`.
+///
+/// Note: bounds aren't generated by this macro,
+/// but you will get compile errors if you do not satisfy them.
+#[manyhow]
+#[proc_macro_derive(ReadConfig, attributes(config))]
+pub fn derive_read_config(input: TokenStream) -> TokenStream {
+    let mut emitter = Emitter::new();
+
+    let Some(input) = emitter.handle(syn::parse2(input)) else {
+        return emitter.finish_token_stream();
+    };
+    let Some(parsed) = emitter.handle(Input::from_derive_input(&input)) else {
+        return emitter.finish_token_stream();
+    };
+    let Some(ir) = parsed.lower(&mut emitter) else {
+        return emitter.finish_token_stream();
+    };
+
+    emitter.finish_token_stream_with(ir.generate())
+}
+
+/// Parsing proc-macro input
+mod ast {
+    use iroha_macro_utils::Emitter;
+    use manyhow::emit;
+    use proc_macro2::{Span, TokenStream, TokenTree};
+    use syn::parse::ParseStream;
+
+    use crate::{codegen, codegen::ParameterEnv};
+
+    // TODO: `attributes(config)` rejects all unknown fields
+    //       it would be better to emit an error "we don't support struct attrs" instead
+    #[derive(darling::FromDeriveInput, Debug)]
+    #[darling(supports(struct_named), attributes(config))]
+    pub struct Input {
+        ident: syn::Ident,
+        generics: syn::Generics,
+        data: darling::ast::Data<(), Field>,
+    }
+
+    impl Input {
+        pub fn lower(self, emitter: &mut Emitter) -> Option<codegen::Ir> {
+            let mut halt_codegen = false;
+
+            for i in self.generics.params {
+                emit!(emitter, i, "generics are not supported");
```

Comment:
> Maybe we can actually add a functionality prefixing the macro name to the emitted error 🤔

I think it would be great. Will see what I can do with `Emitter`.

Reply:
Eeeh, `Emitter` relies on `ToTokensError` trait, which essentially converts everything to a token stream. Therefore, I see the two options:
1. Patch the resulting token stream in a fine-grained manner, prefixing compile error messages. Implications: it might be a complex thing to do?
2. Introduce an intermediate trait for `iroha_macro_utils::Emitter`, which will "intercept" actual errors (plain strings, `syn::Error`, etc) and put a prefix inside them _before_ their conversion to a token stream. Implications: might be a high burden to re-implement the existing thing for a not-so-crucial feature.

What do you think?

For now, I'll simply prefix this particular error.

Reply:
Approach №2 would be cleaner, but that's for a separate PR.